### PR TITLE
second take at setuptools_scm

### PR DESCRIPTION
**Pull request by RonnyPfannschmidt**
since setuptools 18.6 fixes the issues with develop installs

https://github.com/pypa/setuptools/blob/master/CHANGES.rst#186

https://github.com/pypa/setuptools/issues/439

### CODE REVIEW

File: HOWTORELEASE.rst [hunk 1]
```diff
@@ -3,11 +3,9 @@ How to release pytest
 
 Note: this assumes you have already registered on pypi.
 
-1. Bump version numbers in ``_pytest/__init__.py`` (``setup.py`` reads it).
+1. Check and finalize ``CHANGELOG.rst``.
```
Comment (nicoddemus) [inline HOWTORELEASE.rst]: We probably want to mention to generate the tag locally right before creating the package, correct?

**Comment by nicoddemus** [comment]
Besides my comment on `HOWTORELEASE`, linting has fail due to the fact that the numbers in the sections don't increase sequentially anymore.

File: HOWTORELEASE.rst (hunk 1 above)
Comment (RonnyPfannschmidt) [inline HOWTORELEASE.rst]: im considering a setuptools_scm feature where the branch name would be considered as well

**Comment by nicoddemus** [comment]
My previous comment is still valid, I think: 

>We probably want to mention to generate the tag locally right before creating the package, correct?

:grin:

**Comment by RonnyPfannschmidt** [comment]
actually i have a pending setuptools_scm update for solving those details more nice - will show later today

**Comment by nicoddemus** [comment]
Did you give up on this, or will try on another branch?

**Comment by RonnyPfannschmidt** [comment]
@nicoddemus i'll pick this up when either setuptools_scm supports branches or we decide to get rid of the features branch

**Comment by nicoddemus** [comment]
I might be missing something, but I don't really see why. Doesn't setuptools_scm determine the version based on the nearest tag? Só if I tag a commit `2.1.1` the version will be `2.1.1`, and if I decide to create a new feature release I can tag a commit `2.2.0` and it will just work, right?

**Comment by RonnyPfannschmidt** [comment]
@nicoddemus yes, but its a bit of a pain when a feature branch for example has a "wrong" version attached for quite a while

**Comment by nicoddemus** [comment]
Can you elaborate a bit what do you mean by "its a bit of a pain when a feature branch for example has a "wrong" version attached for quite a while"? 

I insist because I would love to use `setuptools_scm` on pytest, and would like to understand the reason completely if we have to ditch it because of having a `features` branch.

**Comment by RonnyPfannschmidt** [comment]
@nicoddemus setuptools_scm currently works based only on tags, so its not aware on how branches should bump the last version, and while it is supported to put certain tags in to help, its not exactly well received to litter the history with extra tags to keep setuptools_scm happy

**Comment by nicoddemus** [comment]
Sorry, I still don't quite get what the issue is, bear with me... from my POV the version is important basically only when making a release, and at that point we will have a tag at the exact location we need it to be. Or am I I missing something?

**Comment by RonnyPfannschmidt** [comment]
if we are fine with such a limitation, we can just switch over right now

**Comment by nicoddemus** [comment]
I'm 👍 on it, unless I'm missing a situation where the development version is important.

**Comment by RonnyPfannschmidt** [comment]
will rebase after pass

**Comment by nicoddemus** [comment]
👍 

Linting broke because of some bad interaction between setuptools-scm and `check-manifest`...

**Comment by nicoddemus** [comment]
@RonnyPfannschmidt I fixed the build locally on Windows:

```diff
diff --git a/_pytest/__init__.py b/_pytest/__init__.py
index 2599bd0..8501718 100644
--- a/_pytest/__init__.py
+++ b/_pytest/__init__.py
@@ -4,4 +4,4 @@ import pkg_resources
 try:
     __version__ = pkg_resources.get_distribution('pytest').version
 except Exception:
-    __version__ = None  # broken installation, we don't even try
+    __version__ = 'unknown'  # broken installation, we don't even try
```

I was about to push to your fork but I got permission denied. 

This brought to light the problem that `pkg_resources.get_distribution` doesn't work on frozen executables (testing-cabal/mock#385). My solution fixes it but leaves the pytest version in the header as "unknown" when pytest is frozen:

```
============================= test session starts =============================
platform win32 -- Python 3.5.0, pytest-unknown, py-1.4.33, pluggy-0.4.0
```  

Do you mind if we use the `write_to` approach instead?

**Comment by RonnyPfannschmidt** [comment]
@nicoddemus the `write_to` approach breaks check-manifest - can we include egg-info into frozen pytest

note that using unknown as version would break expectations as well

**Comment by nicoddemus** [comment]
> can we include egg-info into frozen pytest

Unfortunately no. It think all freeze tools don't support including the packaging metadata (see anthony-tuininga/cx_Freeze#216) and I suspect even after including package metadata `pkg_resources` would still need to learn how to deal with that metadata embedded in zipfiles.

> note that using unknown as version would break expectations as well

I understand, but `None` [breaks pytest](https://ci.appveyor.com/project/pytestbot/pytest/build/1.0.2491/job/qrkf8rshytdmv559) completely...  `"unknown"` as version would break some expectations, but I think it preferable than the former.

> approach breaks check-manifest

I would prefer if we found a solution to that, given that `check-manifest` is a tool we use internally, as opposed to freezing which is something normal users use.

Just to add some context, we freeze pytest extensively together with our applications at work, so we can run the tests using the frozen executable. This is invaluable to catch packaging errors.

File: tox.ini [hunk 2]
```diff
@@ -47,6 +47,9 @@ commands=
 
 [testenv:linting]
 basepython = python2.7
+# needed to keep check-manifest working
+setenv =
+    SETUPTOOLS_SCM_PRETEND_VERSION=2.0.1
```
Comment (nicoddemus) [inline tox.ini]: So it seems this fixes the check-manifest issue. 😁 

Can we switch to the `write_to` option now?

**Comment by RonnyPfannschmidt** [comment]
Nppe, since write_to breaks check manifest

**Comment by nicoddemus** [comment]
> Nppe, since write_to breaks check manifest

I'm a little confused then, why do you need to set `SETUPTOOLS_SCM_PRETEND_VERSION` during linting?

**Comment by RonnyPfannschmidt** [comment]
@nicoddemus pretend_version is for the breakage that is incurred from not having version data in whatever check-manifest copies around, the other breakage is incurred because check-manifest does nit allow for generated files in a sdist that aren't in the scm

**Comment by nicoddemus** [comment]
Perhaps we can use the `ignore` option to ignore the generated version file?

```ini
[check-manifest]
ignore =
    .travis.yml
```

https://pypi.python.org/pypi/check-manifest/

**Comment by RonnyPfannschmidt** [comment]
@nicoddemus thanks for that hint

**Comment by RonnyPfannschmidt** [comment]
@nicoddemus unknown breaks hypothesis - im going to go back to None

**Comment by RonnyPfannschmidt** [comment]
@nicoddemus unknown breaks hypothesis - i'm going to go back to None

**Comment by nicoddemus** [comment]
Hmm wait, but this was working before with hypthosis right? I mean, hypothesis should be seeing the correct version obtained from `_version.py`.

But if we do find a reasonable explanation for the above, how about using `0.0.0` then? That's a valid version string but it is clear something is incorrect with the installation. `None` is not better because pytest itself breaks if __version__ is `None`.

**Comment by RonnyPfannschmidt** [comment]
pytest i can directly fix, hypothesis i cant

**Comment by nicoddemus** [comment]
Why you don't like my suggestion of `"0.0.0"`? It is as good as `"unknown"` or `None` I think.

**Comment by RonnyPfannschmidt** [comment]
@nicoddemus unknown doesnt break on bad version checks due to lexical order, 0.0.0 does in any case 
i'd rather have a correct failure than trying to keep stuff working by means of pretense

**Comment by nicoddemus** [comment]
> i'd rather have a correct failure than trying to keep stuff working by means of pretense

Fair enough. 👍

**Comment by RonnyPfannschmidt** [comment]
@nicoddemus as far as i can tell introducing setuptools_scm will warrant a major release due to the cahnge to pytest.__version__ behaviour

**Comment by The-Compiler** [comment]
What changes about it?

**Comment by RonnyPfannschmidt** [comment]
it could be none, and it an include local version components,
as such it will break naive version parsers

**Comment by nicoddemus** [comment]
> it could be none, and it an include local version components,
as such it will break naive version parsers

But that doesn't affect releases, which will always have a `__version__` in the same format as always. It doesn't make sense to me to increase the major version because of a change that doesn't affect the releases at all. In other words, from a user's POV a release `3.1.0` with or without this change is indistinguishable from each other.

Do you agree?

**Comment by nicoddemus** [comment]
Merging as this now looks great, thanks @RonnyPfannschmidt!

**Comment by The-Compiler** [comment]
This breaks hypothesis, and maybe other plugins doing pytest version checking: https://github.com/HypothesisWorks/hypothesis-python/issues/611

I still maintain my opinion that the benefits of saving some 10s of work for bumping the version number (which can trivially be automated, too) is by no means worth the breakage this causes everywhere...

**Comment by nicoddemus** [comment]
> is by no means worth the breakage this causes everywhere...

Definitely I agree, but that was not supposed to happen. Can you provide more details on how it was possible to install pytest and somehow `setuptools_scm` failing to generate a `_version.py` file? Puzzling...

**Comment by The-Compiler** [comment]
Ah, doing another `pip install -e` run (after pulling the setuptools-scm change) fixes it - sorry for not trying this before. This was a develop install done before setuptools-scm was introduced. 

I wanted to run a `pytest --help` to check something before going to bed and then got this 😆

Now I'm curious though - what's this, and why 1.6.0?

```
$ cat .venv/lib/python3.6/site-packages/_pytest_mock_version.py
# coding: utf-8
# file generated by setuptools_scm
# don't change, don't track in version control
version = '1.6.0'
```

**Comment by nicoddemus** [comment]
> Ah, doing another pip install -e run (after pulling the setuptools-scm change) fixes it - sorry for not trying this before. This was a develop install done before setuptools-scm was introduced.

I figured, no worries. This will only happen if one has an old installation and didn't execute a `pip install -e`. Released packages will always have a `_version.py` file so this should not be an issue. 😉

**Pull request merged**